### PR TITLE
add support for GLES on Windows over angle

### PR DESCRIPTION
**Pull request by AlwinEsch**
This changes allow the addon to use on windows and all other OS's
without add of different code.

The request can not enter now. But already created to have an example, where it is needed.

For this to work, the following changes must be included in Kodi:

| Request                               | Why
|-------------------------------|-------------------
| https://github.com/xbmc/xbmc/pull/15609 | To bring OS independent support for Time
| https://github.com/xbmc/xbmc/pull/15610 | To bring the required Angle support on addon
| https://github.com/xbmc/xbmc/pull/15642 | To have on CMake the right `APP_RENDER_SYSTEM` value set

After Kodi himself has these changes, this request could also come in.

Here a test Picture:
juhu-picture2

Here is a minimal addon, where support this way https://github.com/AlwinEsch/screensaver.example.windows.gles (Maybe good to have better overview)

### CODE REVIEW

File: FindOpenGLES.cmake [hunk 1]
```diff
@@ -9,46 +9,60 @@
...
+# global by the kodi-angle package.
 
-if(CORE_PLATFORM_NAME_LC STREQUAL rbpi)
-    set(_brcmprefix brcm)
-endif()
+if(WIN32)
+  find_package(kodi-angle REQUIRED)
+  set(OPENGLES_LIBRARIES kodi::angle::libGLESv2
+                         kodi::angle::libEGL)
+  set(OPENGLES_FOUND ${kodi-angle_FOUND})
```
Comment (Rechi) [inline FindOpenGLES.cmake:23]: This is wrong, use `find_package_handle_standard_args`.

**Comment by afedchin** [comment]
current CMakeLists.txt for angle is wrong, build will fail for uwp

File: FindOpenGLES.cmake (hunk 1 above)
Comment (AlwinEsch) [reply FindOpenGLES.cmake:23]: Have you an example, has tried now and also on begin and not becomed it working with `find_package_handle_standard_args`?

**Comment by AlwinEsch** [comment]
@afedchin have you a log?

**Comment by afedchin** [comment]
I haven't the log, but I'm sure win32 source shouldn't be included in uwp build.

```
libangle_d3d11_win32_sources = [
  "src/libANGLE/renderer/d3d/d3d11/win32/NativeWindow11Win32.cpp",
  "src/libANGLE/renderer/d3d/d3d11/win32/NativeWindow11Win32.h",
  "src/libANGLE/renderer/d3d/d3d11/converged/CompositorNativeWindow11.cpp",
  "src/libANGLE/renderer/d3d/d3d11/converged/CompositorNativeWindow11.h",
  "src/third_party/systeminfo/SystemInfo.cpp",
  "src/third_party/systeminfo/SystemInfo.h",
]
```
`CompositorNativeWindow11` definitely contains undefined symbols on uwp

File: depends/windows/angle/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+    add_library(angle::renderer::win32 ALIAS angle_renderer_win32)
+    list(APPEND ANGLE_DEPEND_LIBRARIES d3d9.lib)
+  endif()
+
+  ## D3D11 Base renderer
+  if(ANGLE_ENABLE_D3D11)
+    add_library(angle_renderer_d3d INTERFACE)
+    target_sources(angle_renderer_d3d INTERFACE $<BUILD_INTERFACE:${libangle_d3d_shared_sources}>
+                                                $<BUILD_INTERFACE:${libangle_d3d11_sources}>
+                                                $<BUILD_INTERFACE:${libangle_d3d11_win32_sources}>
```
Comment (afedchin) [inline depends/windows/angle/CMakeLists.txt]: `libangle_d3d11_win32_sources` must be excluded for UWP build

**Comment by AlwinEsch** [comment]
Is updated and now works also on Kodi for WindowsStore

File: depends/windows/angle/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+                         -DANGLE_UTIL_EXPORT=
+                         -DEGLAPI=
+                         -DGL_APICALL=
+                         -DGL_API=)
+
+set(GL_PROTOTYPES -DGL_GLES_PROTOTYPES=1
+                  -DEGL_EGL_PROTOTYPES=1
+                  -DGL_GLEXT_PROTOTYPES
+                  -DEGL_EGLEXT_PROTOTYPES
+                  -DHAS_GLES=3
```
Comment (afedchin) [inline depends/windows/angle/CMakeLists.txt]: I think what `HAS_GLES` shouldn't be a part of angle interface.

File: depends/windows/angle/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+if(ANGLE_ENABLE_NULL)
+  add_library(angle_renderer_null INTERFACE)
+  target_sources(angle_renderer_null INTERFACE $<BUILD_INTERFACE:${libangle_gl_null_sources}>)
+  add_library(angle::renderer::null ALIAS angle_renderer_null)
+  list(APPEND LIBANGLE_RENDERER_PLATFORM $<BUILD_INTERFACE:angle::renderer::null>)
+endif()
+
+add_library(libANGLE INTERFACE)
+target_sources(libANGLE INTERFACE $<BUILD_INTERFACE:${libangle_sources}>
+                                  $<BUILD_INTERFACE:${libangle_null_sources}>
```
Comment (afedchin) [inline depends/windows/angle/CMakeLists.txt]: `libangle_null_sources` must be a part of `angle_renderer_null` lib.

File: depends/windows/angle/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+    target_compile_definitions(angle_renderer_win32 INTERFACE -DANGLE_ENABLE_D3D9)
+    target_link_libraries(angle_renderer_win32 INTERFACE d3d9.lib)
+    add_library(angle::renderer::win32 ALIAS angle_renderer_win32)
+    list(APPEND ANGLE_DEPEND_LIBRARIES d3d9.lib)
+  endif()
+
+  ## D3D11 Base renderer
+  if(ANGLE_ENABLE_D3D11)
+    add_library(angle_renderer_d3d INTERFACE)
+    target_sources(angle_renderer_d3d INTERFACE $<BUILD_INTERFACE:${libangle_d3d_shared_sources}>
```
Comment (afedchin) [inline depends/windows/angle/CMakeLists.txt]: as I already said. `libangle_d3d_shared_sources` are common for d3d11 and d3d9. if ANGLE_ENABLE_D3D11=OFF but ANGLE_ENABLE_D3D9=ON then build will fail.

File: depends/windows/angle/CMakeLists.txt (hunk 3 above)
Comment (AlwinEsch) [reply depends/windows/angle/CMakeLists.txt]: Moved back to FindGLES.cmake.

**Comment by AlwinEsch** [comment]
Angles CMakeLists.txt is updated and cleaned up.

Furthermore, I tested it under Linux and ran there after a few changes. The addon in GLES with angle forced and operated with Kodi.

About set to enable or disable d3d11 usage. Unfortunately it is not possible without code changes in angle since he uses a d3d11 header in a global part.

**Comment by afedchin** [comment]
> About set to enable or disable d3d11 usage. Unfortunately it is not possible without code changes in angle since he uses a d3d11 header in a global part.

it's an issue of angle itself otherwise why for introduce `ANGLE_ENABLE_D3D11` definition? and the issue is not a reason to introduce wrong behavior.

**Pull request merged**